### ISSUE TITLE
Questions about crypten and torch training the same model but with different results

### ISSUE THREAD

**Issue by zewen-s** [facebookresearch/CrypTen#461]
I trained and tested two models, Model A and Model B, on the CIFAR10 dataset using both Crypten and PyTorch. The training and testing codes for both models were identical except for the models themselves, and all the experiments were run on a single TitanXP GPU. I used Crypten version 0.4.0 and PyTorch version 1.10.1, with world_size set to 3 for Crypten training and testing.

Model A is the same model used in the Crypten example available on GitHub at https://github.com/facebookresearch/CrypTen/blob/main/examples/mpc_cifar/mpc_cifar.py.

The model is as follows:

```
class LeNet(nn.Sequential):

    def __init__(self):

        super(LeNet, self).__init__()

        self.conv1 = nn.Conv2d(3, 6, 5)

        self.pool = nn.MaxPool2d(2, 2)

        self.conv2 = nn.Conv2d(6, 16, 5)

        self.fc1 = nn.Linear(16 * 5 * 5, 120)

        self.fc2 = nn.Linear(120, 84)

        self.fc3 = nn.Linear(84, 10)

    def forward(self, x):

        x = self.pool(F.relu(self.conv1(x)))

        x = self.pool(F.relu(self.conv2(x)))

        x = x.view(-1, 16 * 5 * 5)

        x = F.relu(self.fc1(x))

        x = F.relu(self.fc2(x))

        x = self.fc3(x)

        return x

```
Model B is a VGG_FeatureExtractor module, defined as follows:

```
class VGG_FeatureExtractor(nn.Module):


    def __init__(self):

        super(VGG_FeatureExtractor, self).__init__()

 

        self.ConvNet = nn.Sequential(

            nn.Conv2d(3, 8, 3, 1, 1), nn.ReLU(True),

            nn.MaxPool2d(2, 2), 

            nn.Conv2d(8, 16, 3, 1, 1), nn.ReLU(True),

            nn.MaxPool2d(2, 2), 

            nn.Conv2d(16, 32, 3, 1, 1), nn.ReLU(True), 

            nn.Conv2d(32, 32, 3, 1, 1), nn.ReLU(True),

            nn.MaxPool2d(2,2),

            nn.Conv2d(32, 64, 3, 1, 1, bias=False),

            nn.BatchNorm2d(64), nn.ReLU(True),

            nn.Conv2d(64, 64, 3, 1, 1, bias=False),

            nn.BatchNorm2d(64), nn.ReLU(True),

            nn.MaxPool2d(2,2), 

            nn.Conv2d(64, 64, 2, 1, 0), nn.ReLU(True))

        self.fc = nn.Linear(64, 10)

    def forward(self, input):

        x = self.ConvNet(input)

        x = x.squeeze(3)      

        x = x.squeeze(2)

        x = self.fc (x)

        return x
```

The result is as follows：

modelA：

![A_t](https://user-images.githubusercontent.com/80888328/231153282-ebc83127-c513-4e88-a28f-85b209706c14.jpg)
![A_i](https://user-images.githubusercontent.com/80888328/231153324-87a18cdd-f47e-40e4-91db-c20e56f3df3c.jpg)

modelB:
![B_t](https://user-images.githubusercontent.com/80888328/231153369-754e522b-c821-420a-a8cf-71042c9a5a58.jpg)
![B_i](https://user-images.githubusercontent.com/80888328/231153393-b8a00458-d942-4659-833f-4de615d8aad7.jpg)

The results of the experiments are shown in the attached image. The results show that in Model A, the difference in accuracy between Crypten and PyTorch on both training and testing sets is very small. However, in Model B, the difference in accuracy between Crypten and PyTorch on both sets is 10 to 20 percentage points. Therefore, I would like to ask if it is normal for the accuracy to differ significantly in Model B, and if not, can you please guide me on what could be causing this? Does the complexity of the model affect the accuracy of Crypten?

In addition, during the experiment, I found that during training or testing, the output value of the crypten model with a small probability will reach the order of 10^9. What is the reason for this?

**Comment by lvdmaaten** [comment]
The main difference between the two models appears to be the use of batch normalization. So perhaps it's related to that? In general, it is important to keep in mind that functions like division and comparators (used for max-pooling and ReLU) are computed using approximations. The numerical errors in these approximations can accumulate. Also, keep in mind that computations are performed using a 16-bit fixed-point number representation rather than a 32-bit floating-point number representations. These things can lead to training instabilities. Therefore, you may need to tune your learning rate (schedule), _etc._ differently in order to obtain good results.

**Comment by zewen-s** [comment]
Thank you very much for your answer and  it helped me a lot.
